Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs

```diff
@@ -142,37 +168,89 @@ public override void OnActionReceived(float[] vectorAction)
+    }
+    
     void FixedUpdate()
     {
+        if (detectTargets)
+        {
+            foreach (var bodyPart in m_JdController.bodyPartsDict.Values)
+            {
+                if (bodyPart.targetContact && bodyPart.targetContact.touchingTarget)
+                {
+                    TouchedTarget();
+                }
+            }
+        }
+        
+        UpdateOrientationCube();
+        
         // Set reward for this step according to mixture of the following elements.
         // a. Velocity alignment with goal direction.
         // b. Rotation alignment with goal direction.
         // c. Encourage head height.
-        // d. Discourage head movement.
-        m_DirToTarget = target.position - m_JdController.bodyPartsDict[hips].rb.position;
         AddReward(
-            +0.03f * Vector3.Dot(m_DirToTarget.normalized, m_JdController.bodyPartsDict[hips].rb.velocity)
-            + 0.01f * Vector3.Dot(m_DirToTarget.normalized, hips.forward)
-            + 0.02f * (head.position.y - hips.position.y)
-            - 0.01f * Vector3.Distance(m_JdController.bodyPartsDict[head].rb.velocity,
-                m_JdController.bodyPartsDict[hips].rb.velocity)
+            +0.02f * Vector3.Dot(m_OrientationCube.transform.forward,
+                Vector3.ClampMagnitude(m_JdController.bodyPartsDict[hips].rb.velocity, maximumWalkingSpeed))
+            + 0.01f * Vector3.Dot(m_OrientationCube.transform.forward, head.forward)
+            + 0.005f * (head.position.y - footL.position.y)
+            + 0.005f * (head.position.y - footR.position.y)
         );
     }
 
     /// <summary>
-    /// Loop over body parts and reset them to initial conditions.
+    /// Agent touched the target
     /// </summary>
-    public override void OnEpisodeBegin()
+    public void TouchedTarget()
     {
-        if (m_DirToTarget != Vector3.zero)
+        AddReward(1f);
+        if (respawnTargetWhenTouched)
         {
-            transform.rotation = Quaternion.LookRotation(m_DirToTarget);
+            GetRandomTargetPos();
         }
+    }
+
+    /// <summary>
+    /// Moves target to a random position within specified radius.
+    /// </summary>
+    public void GetRandomTargetPos()
+    {
+        var newTargetPos = Random.insideUnitSphere * targetSpawnRadius;
+        newTargetPos.y = 5;
+        target.position = newTargetPos + ground.position;
+    }
 
+    /// <summary>
+    /// Loop over body parts and reset them to initial conditions.
+    /// </summary>
+    public override void OnEpisodeBegin()
+    {
+        //Reset all of the body parts
         foreach (var bodyPart in m_JdController.bodyPartsDict.Values)
         {
             bodyPart.Reset(bodyPart);
         }
+        
+        //Random start rotation to help generalize
+        transform.rotation = Quaternion.Euler(0, Random.Range(0.0f, 360.0f), 0);
+        
+        UpdateOrientationCube();
+
+        if (detectTargets && !targetIsStatic)
```

Comment:
Not too sure what `detectTargets && !targetIsStatic` does. Can you add a comment for reference.

Reply:
this is logic related to respawning the target. another artifact from not having a dedicated target controller script.

Reply:
added code comment